otherlibs/stdune/src/ansi_color.ml

```diff
@@ -316,20 +386,61 @@ let parse_line str styles =
                  the same as "\027[0m" by terminals *)
               []
             else
+              let style_of_string accu s =
+                match Int.of_string s with
+                | None -> accu
+                | Some code -> (
+                  match Style.of_ansi_code code with
+                  | `Clear -> []
+                  | `Unknown -> accu
+                  (* If the foreground is set to default, we filter out any
+                     other foreground modifiers. Same for background. *)
+                  | `Fg_default -> List.filter accu ~f:Style.is_not_fg
+                  | `Bg_default -> List.filter accu ~f:Style.is_not_bg
+                  | #Style.t as s -> s :: accu)
+              in
+              let rec parse_styles (accu : Style.t list) l =
+                match l with
+                | [] -> accu
+                (* Parsing 8-bit foreground colors *)
+                | "38" :: "5" :: s :: l -> (
```

Comment:
I think this is getting ugly enough that we should probably pull out ocamllex to do this work.